bauble/src/value/display.rs

```diff
@@ -0,0 +1,593 @@
+            }
+            Value::Struct(fields_kind) => {
+                w.write_type();
+                match fields_kind {
+                    FieldsKind::Unit => {}
+                    FieldsKind::Unnamed(items) => {
+                        w.write("(");
+                        slice_display(items, w.reborrow());
+                        w.write(")");
+                    }
+                    FieldsKind::Named(items) => {
+                        w.write(" {");
+                        if !items.is_empty() {
+                            w.write_recursive(|mut f| {
+                                for (field, value) in items {
+                                    f.write_line(|mut l| {
+                                        l.write(field.as_str());
+                                        l.write(": ");
+                                        value.indented_display(l.reborrow());
+                                        l.write(",");
+                                    });
+                                }
+                            });
+                        }
+                        w.write("}");
+                    }
+                }
+            }
+            Value::Or(or) => match or.as_slice() {
+                [] => w.write("|"),
+                [variant] => {
+                    w.write_type_path();
+                    w.fmt(variant);
+                }
+                [rest @ .., last] => {
+                    for v in rest {
+                        w.write_type_path();
+                        w.fmt(v);
+                        w.write(" | ");
+                    }
+                    w.fmt(last);
+                }
+            },
+            Value::Primitive(primitive_value) => match primitive_value {
+                super::PrimitiveValue::Num(v) => w.fmt(v),
+                super::PrimitiveValue::Str(v) => w.debug_fmt(v),
+                super::PrimitiveValue::Bool(v) => w.fmt(v),
+                super::PrimitiveValue::Unit => w.write("()"),
+                // `Raw` is a bit annoying to handle here, since we don't know how it
+                // was originally expressed. So we need to check:
+                // 1. Can it be displayed as a literal raw, i.e `#some_raw`
+                // 2. If not, how many braces do we need, since raw values can contain
+                //    `}`, but it has to end with as many `}` as there was `{` at the
+                //    start.
+                super::PrimitiveValue::Raw(v) => {
+                    let mut num_braces = None;
+                    let mut row = None;
+
+                    let mut first = None;
+                    let mut last = None;
+                    for c in v.chars() {
+                        if first.is_none() {
+                            first = Some(c);
+                        }
+                        last = Some(c);
+                        if num_braces.is_none() && !allowed_in_raw_literal(c) {
```

Comment:
this one is a bit confusing, because are you not specifically looking for a starting brace `{`, but here you're matching with any non-valid string character which looking at `allowed_in_raw_literal` is more than just braces?

Reply:
If there are characters that aren't allowed in a raw literal we need to put braces yea, which includes `{`. But we only need to add more than 1 brace if the raw contains `}`